### ISSUE TITLE
LazyTrait and apply_transform API issue

### ISSUE THREAD

**Issue by wyli** [Project-MONAI/MONAI#6597]
**Describe the bug**
The pattern of invoking transforms with a `lazy` parameter depending on `isinstance(transform, LazyTrait)` is problematic
https://github.com/Project-MONAI/MONAI/blob/e6ec945e4b87b90835cdad29ea64b1f27b8accda/monai/transforms/transform.py#L98
because the `LazyTrait` API doesn't provide a `__call__` design.
https://github.com/Project-MONAI/MONAI/blob/e6ec945e4b87b90835cdad29ea64b1f27b8accda/monai/transforms/traits.py#L22-L24
 
(introduced since https://github.com/Project-MONAI/MONAI/pull/6537)

**Comment by atbenmurray** [comment]
@wyli I should be part of this discussion.
@mjorgecardoso

**Comment by wyli** [comment]
> @wyli I should be part of this discussion.
> @mjorgecardoso


Hi Ben, there has been concerns raised from the core team including myself, that contributor doesn't accept any code design suggestions from the core dev team. You can join this discussion if you agree that we can improve this together for efficient discussions.

**Comment by mjorgecardoso** [comment]
Hi @wyli, please keep @atbenmurray involved in all conversation involving the Lazy transforms as we all agreed during the MONAI meetings. His concerns regarding the architectural decisions have merit and should be taken into account.

**Comment by wyli** [comment]
@mjorgecardoso the core dev team is open to ideas and criticisms but we also emphasize healthy discussions which requires the contributors being collaborative. Cc @ericspod @mmodat @aylward @anfeng

**Comment by mjorgecardoso** [comment]
Of course. But please be courteous and ensure the feature lead is involved in all discussions involving that feature.

**Comment by wyli** [comment]
v1.2 has been released and milestones are closed. we are currently defining v1.3, the feature lead has been defined yet.

**Comment by atbenmurray** [comment]
> @mjorgecardoso the core dev team is open to ideas and criticisms but we also emphasize healthy discussions which requires the contributors being collaborative. Cc @ericspod @mmodat @aylward @anfeng

Yes, this has been a critical issue. If I were to enumerate my concerns:
- The design of lazy resampling was completely changed by you without any discussion with me
- You merged it to dev while I was still reviewing it. You said "Nic reviewed it" but it was obvious that I should also have signed off on it
- The resulting design broke pipelines in the general case
- I agreed to work with you to duplicate the "out_of_order" pipeline behaviour to avoid undoing the work of the wider team
- You raised many concerns, some of which I agreed with and some of which I thought were incorrect
- At the point you couldn't argue from a technical standpoint, you cited fatigue and then created a new PR that evening to cut out my work
- When the consensus in the core dev meeting didn't agree with you, you completely disengaged from the process rather than continuing to work with me on a suitable compromise
- The remaining process took place without your input despite being invited to review subsequent
  - I took Eric and Nic and Vikash through the design with a fine-toothed comb to ensure that they were happy with it
  - I worked through a comprehensive review process with Eric and Nic to resolve their concerns
  - When it turned out that everyone was struggling with understanding the out of order behaviour I wrote a comprehensive design document for lazy resampling and solicited feedback
  - When I asked you why you had changed your mind about "out_of_order" resampling, you said "not your implementation of it"
  - You have loudly and regularly stated that I have delayed the release. What actually delayed the release was the design of lazy resampling being changed to a design that would make most pipelines fail / output unexpected results and then committing it to dev while I was still reviewing it
- You have undermined me in meetings on a regular basis, including calling me a liar in the most recent meeting
- I have remained (and continue to remain) courteous and professional throughout

Please also remember that I am actually a member of the core dev team

**Comment by wyli** [comment]
> > @wyli I should be part of this discussion.
> > @mjorgecardoso
> 
> Hi Ben, there has been concerns raised from the core team including myself, that contributor doesn't accept any code design suggestions from the core dev team.

Please note that the concerns are raised from the core dev team members about Ben, it's not just my opinion...

**Comment by atbenmurray** [comment]
This can be very simply disproved by actually looking at the most recent PRs.

**Comment by mjorgecardoso** [comment]
Hi all, let’s meet and chat about this. Let’s make it constructive.

**Comment by wyli** [comment]
Hi @mjorgecardoso, all my comments here are either describing the facts or being constructive. I'd recommend you align with your team internally first before we chat...

**Comment by atbenmurray** [comment]
I'm hoping that we can continue with the implementation of lazy resampling for 1.4. We should have a discussion about the use of interfaces and the design choices around them as part of this.
